WcaOnRails/app/controllers/api/v0/competitions_controller.rb

```diff
@@ -1,11 +1,69 @@
 # frozen_string_literal: true
 class Api::V0::CompetitionsController < Api::V0::ApiController
-  def show
-    competition = Competition.where(id: params[:id], showAtAll: true).first
-    unless competition
-      render json: { error: "Competition with id #{params[:id]} not found" }, status: 404
-      return
+  def index
+    managed_by_user = nil
+    if params[:managed_by_me].present?
+      require_scope!("manage_competitions")
+      managed_by_user = current_resource_owner
     end
+
+    competitions = Competition.search(params[:q], params: params, managed_by_user: managed_by_user)
+    competitions = competitions.includes(:delegates, :organizers)
+
+    paginate json: competitions
+  end
+
+  def show
+    competition = competition_from_params
     render json: competition
   end
+
+  def big_json
+    competition = competition_from_params
+    require_can_manage!(competition)
+
+    render json: competition.competition_big_json.big_json
+  end
+
+  def update_big_json
+    competition = competition_from_params
+    require_can_manage!(competition)
+
+    new_big_json = params.require(:big_json).permit!.to_h
+    comp_big_json = competition.competition_big_json
+    comp_big_json.big_json = new_big_json
+    if comp_big_json.save
+      render json: comp_big_json.big_json
+    else
+      render status: 400, json: { error: competition_big_json.errors.full_messages.join("\n") }
+    end
+  end
+
+  private def competition_from_params
+    id = params[:competition_id] || params[:id]
+    competition = Competition.find_by_id(id)
+
+    # If this competition exists, but is not publicly visible, then only show it
+    # to the user if they are able to manage the competition.
+    if competition && !competition.showAtAll && !can_manage?(competition)
+      competition = nil
+    end
+
+    raise WcaExceptions::NotFound.new("Competition with id #{id} not found") unless competition
```

Comment:
For the record, if this ever *does* cause an issue, I am happy to revert my stance on "we shouldn't leak the ids of upcoming competitions". It's just something that the original PHP code seemed to do, and so I've been trying to maintain that constraint. I don't think it's actually that important, however, and no one from other parts of the WCA has told me this is a very important task.